Throw type error on converting non objects to sequence

stspyder opened the pull request:
Fix for #237

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/rust-mozjs/238)
<!-- Reviewable:end -->

KiChjang commented on src/conversions.rs:
```diff
@@ -508,7 +509,10 @@ impl<C: Clone, T: FromJSValConvertible<Config=C>> FromJSValConvertible for Vec<T
                 let mut val = RootedValue::new(cx, UndefinedValue());
-                assert!(JS_GetElement(cx, obj.handle(), i, val.handle_mut()));
+                if !JS_GetElement(cx, obj.handle(), i, val.handle_mut()) {
+                    // On JS exception or error exit loop
```
So no panic nor any error thrown?

stspyder replied:
Doesn't 'JS_GetElement' automatically report JS exceptions? I thought so because the tests with assert_throws(custom_error, fun()) were passing.

KiChjang replied:
Where is this assert_throws(custom_error, fun()) test?

KiChjang replied:
As far as I can tell, JS_GetElement is an external C function, and it doesn't use any of the `throw_js_error` stuff that we have written in Rust.

stspyder replied:
Well, I meant the WPT tests. I thought I saw the automatic exception reporting stuff somewhere in the SpiderMonkey documentation. May be it doesn't. I'll see if I can test.

Ms2ger replied:
I thought this couldn't fail, but I double-checked, and it can:
```js
Object.defineProperty(Array.prototype, 1, {get: function() {throw 7}});
new Blob([,,,])
```

stspyder commented:
I have updated the pull request. From [JSAPI guide](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/JSAPI_User_Guide#Automatic_handling_of_uncaught_exceptions), it seems that JSAPI functions automatically sets the exception (thrown from JS) in the context and return false without we having to explicitly set one in the context. This exception will be cleared eventually down the stack.

I have updated the code to return an Err in case of a JS exception so that the constructor can return false if the conversion fails.

Ms2ger commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit f2f0770 has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger f2f07700e233daf07be2c1d9fdf943f0c1b7bb89 -->

bors-servo commented:
:hourglass: Testing commit f2f0770 with merge 62d23de...

bors-servo commented:
:sunny: Test successful - [travis](https://travis-ci.org/servo/rust-mozjs/builds/107817953)
